### PR TITLE
feat: add tx disable FAST confirmation

### PR DESCRIPTION
**Pull request by ws4charlie**
# Description

1. add `MsgDisableFastConfirmation` message to disable `FAST` confirmation for given `chain_id`
2. add migration script to set `MsgDisableFastConfirmation` under emergency policy.

Closes: https://github.com/zeta-chain/node/issues/3462

# How Has This Been Tested?

- [ ] Tested CCTX in localnet
- [ ] Tested in development environment
- [x] Go unit tests
- [ ] Go integration tests
- [ ] Tested via GitHub Actions

## Summary by CodeRabbit

- **New Features**
  - Introduced a new option that allows users to disable fast confirmation for selected blockchain networks via a dedicated CLI command.
  - Added support for tracking cross-chain transactions and enhanced Sui chain information availability.
- **Documentation**
  - Updated command guides and API references with detailed usage instructions for the new fast confirmation disabling feature.
- **Bug Fixes**
  - Improved confirmation processing and error handling to boost transaction reliability.

### CODE REVIEW

File: proto/zetachain/zetacore/observer/tx.proto [hunk 1]
```diff
@@ -162,3 +164,12 @@ message MsgUpdateOperationalFlags {
 }
 
 message MsgUpdateOperationalFlagsResponse {}
+
+// MsgDisableFastConfirmation is used to disable fast confirmation for a list of
+// chains.
+message MsgDisableFastConfirmation {
+  option (cosmos.msg.v1.signer) = "creator";
+  string creator = 1;
+  repeated int64 chain_id_list = 2;
```
Comment (lumtis) [inline proto/zetachain/zetacore/observer/tx.proto:173]: Easier to have a single chain instead

We can bundle several messages in the same tx. This would make the state transition more atomic and simpler to test

**Review by lumtis — CHANGES_REQUESTED** [review]

File: proto/zetachain/zetacore/observer/tx.proto (hunk 1 above)
Comment (ws4charlie) [reply proto/zetachain/zetacore/observer/tx.proto:173]: @lumtis to disable all, passing `chain_id == 0` looks good to you? because it's during emergency.

Comment (lumtis) [reply proto/zetachain/zetacore/observer/tx.proto:173]: If we are concerned about a emergency scenario we can prepare in advance a proposal that include all chains.
I practice I think it will be more case by case?
If we want to keep this option then better to add a bool `all_chains` to make things explicit

Comment (swift1337) [reply proto/zetachain/zetacore/observer/tx.proto:173]: I agree that `chain_id` is simpler. chaid_id == 0 is opaque. Maybe `chain_id = -1` could indicate "all chains"

Comment (swift1337) [reply proto/zetachain/zetacore/observer/tx.proto:173]: Also, how to enable it back later? Should we use `MsgSetFastConfirmation(chain_id int64, enabled bool)`?

Comment (ws4charlie) [reply proto/zetachain/zetacore/observer/tx.proto:173]: okay, let me just keep this simplest case. Only takes one `chain_id` and operation team has a prepared proposal at hand to disable all chains.

Comment (ws4charlie) [reply proto/zetachain/zetacore/observer/tx.proto:173]: To bring it back, it goes to the `MsgUpdateChainParamsResponse` under `admin` policy.

Comment (ws4charlie) [reply proto/zetachain/zetacore/observer/tx.proto]: changed `chain_id_list` -> `chain_id` https://github.com/zeta-chain/node/pull/3522/commits/4d4e040cef64cc18f928cada5e9e659b89bf7723

File: x/observer/keeper/msg_server_disable_fast_confirmation.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+	// find current chain params list
+	chainParamsList, found := k.GetChainParamsList(ctx)
+	if !found {
+		return &types.MsgDisableFastConfirmationResponse{}, types.ErrChainParamsNotFound
+	}
+
+	// disable fast confirmation by setting fast confirmation count to zero
+	for _, cp := range chainParamsList.ChainParams {
+		if cp.ConfirmationParams == nil {
+			continue // should never happen
```
Comment (lumtis) [inline x/observer/keeper/msg_server_disable_fast_confirmation.go:36]: Can we instead check only for the chain matching the chain ID and return an error when it is nil

Comment (lumtis) [inline x/observer/keeper/msg_server_disable_fast_confirmation.go:36]: Let's return an error if the chain is not found

File: x/observer/keeper/msg_server_disable_fast_confirmation.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+		if cp.ChainId == msg.ChainId {
+			cp.ConfirmationParams.FastInboundCount = cp.ConfirmationParams.SafeInboundCount
+			cp.ConfirmationParams.FastOutboundCount = cp.ConfirmationParams.SafeOutboundCount
+		}
+	}
+
+	// validate the updated chain params list
+	if err := chainParamsList.Validate(); err != nil {
+		return &types.MsgDisableFastConfirmationResponse{}, errorsmod.Wrap(types.ErrInvalidChainParams, err.Error())
+	}
```
Comment (lumtis) [inline x/observer/keeper/msg_server_disable_fast_confirmation.go:50]: Let's remove this

If tehre is an edge case where another data in the chain param is invlid, this message would be blocked while we need to run it as an emergency

The test should check that this state transisition doesn't create an invlaid chain param list

File: x/observer/keeper/msg_server_disable_fast_confirmation.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+	// check permission
+	err := k.GetAuthorityKeeper().CheckAuthorization(ctx, msg)
+	if err != nil {
+		return nil, errorsmod.Wrap(authoritytypes.ErrUnauthorized, err.Error())
+	}
+
+	// find current chain params list
+	chainParamsList, found := k.GetChainParamsList(ctx)
+	if !found {
+		return &types.MsgDisableFastConfirmationResponse{}, types.ErrChainParamsNotFound
```
Comment (lumtis) [inline x/observer/keeper/msg_server_disable_fast_confirmation.go:30]: Inconsistent, sometime we return nil and sometimes an empty value
Let's return nil everywhere

File: x/observer/keeper/msg_server_disable_fast_confirmation.go [hunk 5]
```diff
@@ -0,0 +1,56 @@
+package keeper
+
+import (
+	"context"
+
+	errorsmod "cosmossdk.io/errors"
```
Comment (lumtis) [inline x/observer/keeper/msg_server_disable_fast_confirmation.go:6]:
```suggestion
	cosmoserror "cosmossdk.io/errors"
```

File: x/observer/keeper/msg_server_disable_fast_confirmation_test.go [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+		require.Len(t, chainParamsList.ChainParams, 1)
+		require.Equal(t, chainParams, chainParamsList.ChainParams[0])
+	})
+}
+
+// ensureFastConfirmationEnabled checks that the fast confirmation is enabled
+func ensureFastConfirmationEnabled(t *testing.T, chainParams *types.ChainParams) {
+	require.NotNil(t, chainParams.ConfirmationParams)
+	require.Positive(t, chainParams.ConfirmationParams.FastInboundCount)
+	require.Positive(t, chainParams.ConfirmationParams.FastOutboundCount)
```
Comment (lumtis) [inline x/observer/keeper/msg_server_disable_fast_confirmation_test.go:191]: I think it's not up to date, should check lower than safe

File: x/observer/simulation/operation_disable_fast_confirmation.go [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+	moduletestutil "github.com/cosmos/cosmos-sdk/types/module/testutil"
+	simtypes "github.com/cosmos/cosmos-sdk/types/simulation"
+	"github.com/cosmos/cosmos-sdk/x/simulation"
+
+	zetasimulation "github.com/zeta-chain/node/testutil/simulation"
+	"github.com/zeta-chain/node/x/observer/keeper"
+	"github.com/zeta-chain/node/x/observer/types"
+)
+
+// SimulateDisableFastConfirmation generates a MsgDisableFastConfirmation and delivers it.
```
Comment (lumtis) [inline x/observer/simulation/operation_disable_fast_confirmation.go:17]: Note for future PR: no need to add simulation for new messages

File: x/observer/keeper/msg_server_disable_fast_confirmation.go (hunk 2 above)
Comment (ws4charlie) [reply x/observer/keeper/msg_server_disable_fast_confirmation.go]: return error on nil https://github.com/zeta-chain/node/pull/3522/commits/05ece1e475ab004869bd8297f35127d5d9870f31

Comment (ws4charlie) [reply x/observer/keeper/msg_server_disable_fast_confirmation.go]: return error if no matching chain ID found https://github.com/zeta-chain/node/pull/3522/commits/05ece1e475ab004869bd8297f35127d5d9870f31

File: x/observer/keeper/msg_server_disable_fast_confirmation.go (hunk 3 above)
Comment (ws4charlie) [reply x/observer/keeper/msg_server_disable_fast_confirmation.go]: removed https://github.com/zeta-chain/node/pull/3522/commits/05ece1e475ab004869bd8297f35127d5d9870f31

File: x/observer/keeper/msg_server_disable_fast_confirmation.go (hunk 4 above)
Comment (ws4charlie) [reply x/observer/keeper/msg_server_disable_fast_confirmation.go]: updated in the same commit https://github.com/zeta-chain/node/pull/3522/commits/05ece1e475ab004869bd8297f35127d5d9870f31

File: x/observer/keeper/msg_server_disable_fast_confirmation.go (hunk 5 above)
Comment (ws4charlie) [reply x/observer/keeper/msg_server_disable_fast_confirmation.go]: https://github.com/zeta-chain/node/pull/3522/commits/05ece1e475ab004869bd8297f35127d5d9870f31

File: x/observer/keeper/msg_server_disable_fast_confirmation_test.go (hunk 6 above)
Comment (ws4charlie) [reply x/observer/keeper/msg_server_disable_fast_confirmation_test.go:220]: adjusted sample function https://github.com/zeta-chain/node/pull/3522/commits/05ece1e475ab004869bd8297f35127d5d9870f31

File: x/observer/simulation/operation_disable_fast_confirmation.go (hunk 7 above)
Comment (ws4charlie) [reply x/observer/simulation/operation_disable_fast_confirmation.go:17]: Noted.

File: proto/zetachain/zetacore/observer/tx.proto (hunk 1 above)
Comment (ws4charlie) [reply proto/zetachain/zetacore/observer/tx.proto]: I'll mark this as resolved.

File: x/observer/keeper/msg_server_disable_fast_confirmation.go [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+	if !found {
+		return nil, types.ErrChainParamsNotFound
+	}
+
+	// disable fast confirmation
+	foundChain := false
+	for _, cp := range chainParamsList.ChainParams {
+		if cp.ConfirmationParams == nil {
+			return nil, types.ErrInvalidChainParams
+		}
```
Comment (lumtis) [inline x/observer/keeper/msg_server_disable_fast_confirmation.go:38]: Let's add this check below when the chain is found

Comment (ws4charlie) [reply x/observer/keeper/msg_server_disable_fast_confirmation.go:43]: updated in commit https://github.com/zeta-chain/node/pull/3522/commits/8f1370716f1d6714ef82a60dee38fa080adf9786

File: x/observer/keeper/msg_server_disable_fast_confirmation.go [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+		if cp.ConfirmationParams == nil {
+			return nil, types.ErrInvalidChainParams
+		}
+
+		// setting fast confirmation count to same value as safe confirmation count
+		// will effectively disable fast confirmation
+		foundChain = true
+		cp.ConfirmationParams.FastInboundCount = cp.ConfirmationParams.SafeInboundCount
+		cp.ConfirmationParams.FastOutboundCount = cp.ConfirmationParams.SafeOutboundCount
```
Comment (lumtis) [inline x/observer/keeper/msg_server_disable_fast_confirmation.go:49]:
```suggestion
		if cp.ChainId == msg.ChainId {
			if cp.ConfirmationParams == nil {
				return nil, types.ErrInvalidChainParams
			}

			// setting fast confirmation count to same value as safe confirmation count
			// will effectively disable fast confirmation
			foundChain = true
			cp.ConfirmationParams.FastInboundCount = cp.ConfirmationParams.SafeInboundCount
			cp.ConfirmationParams.FastOutboundCount = cp.ConfirmationParams.SafeOutboundCount
			break
		}	
```

I think it's cleaner this way

File: x/observer/keeper/msg_server_disable_fast_confirmation.go [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+	// find current chain params list
+	chainParamsList, found := k.GetChainParamsList(ctx)
+	if !found {
+		return nil, types.ErrChainParamsNotFound
+	}
+
+	// disable fast confirmation
+	foundChain := false
+	for _, cp := range chainParamsList.ChainParams {
+		if cp.ChainId != msg.ChainId {
```
Comment (lumtis) [inline x/observer/keeper/msg_server_disable_fast_confirmation.go:37]: Would be more visual to modify
```
chainParamsList.ChainParams[i]
```
instead of `cp`, unclear `cp` is mutable

File: x/observer/keeper/msg_server_disable_fast_confirmation.go [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+	}
+	if !foundChain {
+		return nil, cosmoserror.Wrap(
+			types.ErrChainParamsNotFound,
+			fmt.Sprintf("no matching chain ID found (%d)", msg.ChainId),
+		)
+	}
+
+	// set the updated chain params list
+	k.SetChainParamsList(ctx, chainParamsList)
```
Comment (lumtis) [inline x/observer/keeper/msg_server_disable_fast_confirmation.go:59]: let's put this in the loop after it's modified

File: x/observer/keeper/msg_server_disable_fast_confirmation.go (hunk 9 above)
Comment (ws4charlie) [reply x/observer/keeper/msg_server_disable_fast_confirmation.go:49]: do you really want the additional break? I didn't add in case we have duplicate chain ids.

Comment (ws4charlie) [reply x/observer/keeper/msg_server_disable_fast_confirmation.go]: updated in commit https://github.com/zeta-chain/node/pull/3522/commits/24d6b53ab91c6f5ef082b3718d14b5076d4a330c

File: x/observer/keeper/msg_server_disable_fast_confirmation.go (hunk 10 above)
Comment (ws4charlie) [reply x/observer/keeper/msg_server_disable_fast_confirmation.go]: same commit https://github.com/zeta-chain/node/pull/3522/commits/24d6b53ab91c6f5ef082b3718d14b5076d4a330c

**Review by lumtis — APPROVED** [review]
